compiler/src/dotty/tools/dotc/core/TypeApplications.scala

```diff
@@ -168,11 +168,16 @@ class TypeApplications(val self: Type) extends AnyVal {
    *  any type parameter that is-rebound by the refinement.
    */
   final def typeParams(implicit ctx: Context): List[TypeParamInfo] = /*>|>*/ track("typeParams") /*<|<*/ {
+    def isTrivial(prefix: Type, tycon: Symbol) = prefix match {
+      case prefix: ThisType => prefix.cls `eq` tycon.owner
+      case NoPrefix => true
+      case _ => false
+    }
     try self match {
       case self: TypeRef =>
         val tsym = self.symbol
         if (tsym.isClass) tsym.typeParams
-        else if (!tsym.exists) self.info.typeParams
+        else if (!tsym.exists || !isTrivial(self.prefix, tsym)) self.info.typeParams
```

Comment:
There will always be cyclic errors. The challenge is to reduce them sufficiently that 

 - they occur only rarely
 - they can be communicated so that they make sense to the user
 - there is a way to work around them

and, probably most importantly

 - they don't break existing code. What worked with the current compiler should continue working.

So it's really quite fuzzy. Ideally, we should have a spec that says in exactly what situations cyclic errors are OK. But we are currently quite far away from this point. I believe most mainstream languages that allow unlimited recursion face the same fundamental problem.